numerics/quadrature_body.hpp

```diff
@@ -55,6 +57,78 @@ Primitive<std::invoke_result_t<Function, Argument>, Argument> GaussLegendre(
                        GaussLegendreWeights[points]);
 }
 
+// Our automatic Cleshaw-Curtis implementation doubles the number of points
+// repeatedly until it reaches a suitable exit criterion.  Naïvely evaluating
+// the function N times for each iteration would be wasteful.  Assume that we
+// have already evaluated the function at 0/2 and 1/2 (that is, s = 0 and 1 for
+// N = 2), we would evaluate it again at 0/4 and 2/4 for N = 4, and then proceed
+// with evaluations at 1/4 and 3/4.  Overall, we would do roughly twice the
+// work.
+// Therefore, we cache the past evaluations.  However, we do not want to use a
+// map and we do not want to do insertion in a vector, say, to squeeze 1/4
+// between 0/2 and 1/2.  So we append the result of evaluations to the cache
+// vector as we perform them.  To fill the cache for N points, we recursively
+// fill it for N/2 points (which yields all the values at s/N for s even) and
+// then we append the values at s/N for s odd.  The order of the second part
+// matters.  It is done by taking the indices of the first part (covering
+// [0/N, (N/2-1)/N[) in bit-reversed order and for each index s appending the
+// value at (2 s + 1)/N.  This ensures that the entire array follows
+// bit-reversed ordering.
+// There is an extra wart: we need the value at N/N, but it does not fit nicely
+// in the bit-reversed ordering.  So we store it at position 0 in the cache, and
+// the regular cache starts at index 1.
+// Clients are expected to reserve (at least) points entries in the cache vector
+// for efficient heap allocation.
+template<int points, typename Argument, typename Function>
+void FillClenshawCurtisCache(
+    Function const& f,
+    Argument const& lower_bound,
+    Argument const& upper_bound,
+    std::vector<std::invoke_result_t<Function, Argument>>& cached_f_cos_N⁻¹π) {
```

Comment:
Rename this to `cached_evaluations`, or alternatively call it `f_cos_N⁻¹π_bit_reversed`.

As it stands it looks like this uses the same convention as `f_cos_N⁻¹π`, that is, `f_cos_N⁻¹π[s]` = f(cos N⁻¹πs), which is not the case.